adapters/adhese/adhese.go

```diff
@@ -0,0 +1,282 @@
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{WrapError(fmt.Sprintf("Unexpected status code: %d.", response.StatusCode))}
+	}
+
+	var originArray []openrtb_ext.AdheseOrigin
+	var bidResponse openrtb.BidResponse
+
+	if err := json.Unmarshal(response.Body, &originArray); err != nil {
+		return nil, []error{err, WrapError(fmt.Sprintf("Response %v does not have an Origin.", string(response.Body)))}
+	}
+
+	if originArray[0].Origin == "JERLICIA" || originArray[0].Origin == "DALE" {
+		var adheseBidResponseArray []openrtb_ext.AdheseBid
+		if err := json.Unmarshal(response.Body, &adheseBidResponseArray); err != nil {
+			return nil, []error{err, WrapError(fmt.Sprintf("Response %v could not be parsed as Adhese bid.", string(response.Body)))}
+		}
+		bidResponse = convertAdheseBid(adheseBidResponseArray[0])
+	} else {
+		var openRtbBidResponseArray []openrtb_ext.AdheseOpenRtbBid
+		if err := json.Unmarshal(response.Body, &openRtbBidResponseArray); err != nil {
+			return nil, []error{err, WrapError(fmt.Sprintf("Response %v could not be parsed as Adhese OpenRtb Bid.", string(response.Body)))}
+		}
+		bidResponse = convertAdheseOpenRtbBid(openRtbBidResponseArray[0])
+	}
+
+	bidderResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+
+	if len(bidResponse.SeatBid) == 0 {
+		return nil, []error{WrapError("Response resulted in an empty seatBid array.")}
+	}
+
+	var errs []error
+	for _, sb := range bidResponse.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+			bid := sb.Bid[i]
+			bidderResponse.Bids = append(bidderResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: getBidType(bid.AdM),
+			})
+
+		}
+	}
+	return bidderResponse, errs
+}
+
+func convertAdheseBid(adheseBid openrtb_ext.AdheseBid) openrtb.BidResponse {
+	price, _ := strconv.ParseFloat(adheseBid.Extension.Prebid.Cpm.Amount, 64)
+	width, _ := strconv.ParseUint(adheseBid.Width, 10, 64)
+	height, _ := strconv.ParseUint(adheseBid.Height, 10, 64)
+	adheseObj, _ := json.Marshal(openrtb_ext.ExtAdhese{
+		CreativeId:                adheseBid.Id,
+		AdFormat:                  adheseBid.AdFormat,
+		AdType:                    adheseBid.AdType,
+		AdspaceId:                 adheseBid.AdspaceId,
+		DealId:                    adheseBid.OrderId,
+		LibId:                     adheseBid.LibId,
+		OrderProperty:             adheseBid.OrderProperty,
+		Priority:                  adheseBid.Priority,
+		ViewableImpressionCounter: adheseBid.ViewableImpressionCounter,
+	})
+	return openrtb.BidResponse{
+		ID: adheseBid.Id,
+		SeatBid: []openrtb.SeatBid{{
+			Bid: []openrtb.Bid{{
+				ID:     adheseBid.Id,
+				ImpID:  adheseBid.CreativeName,
+				DealID: adheseBid.OrderId,
+				Price:  price,
+				W:      width,
+				H:      height,
+				CID:    adheseBid.OrderId,
+				CrID:   adheseBid.Id,
+				NURL:   adheseBid.ImpressionCounter,
```

Comment:
This is wrong - either we output a win notice url and expect someone to call it, or we append it in the adMarkup
At the moment it seems we do both.

Reply:
So we prefer the admarkup because that goes without relying on someone else calling it?

Reply:
That's whad the js prebid seems to be doing... so maybe we should stick with the same logic?